File: apps/web/lib/workflow-planner/assemble.ts
// From a problem to a workflow: a walk down the layer graph's containment.
//
// A capability is realised by one of its methods; a method needs a list of
// smaller capabilities (`steps`); each of those is realised by one of ITS
// methods; and so on. A workflow is one choice at every capability on that
// walk. The planner makes a default choice and says why, and the reader can
// replace any choice with any other method that realises the same capability —
// the "lego" the owner asked the Atlas to be (block-repository roadmap §0.1).
//
// ## How a default is chosen, in order, and what the page says about it
//
// 1. `reader` — the reader picked it.
// 2. `published` — the parent method's own `via` names the method its source
//    used for this step. This is the only choice with a paper behind it.
// 3. `preferred` — the problem's `preferredMethods` (root only), a planner
//    ordering stated in `./problems.ts`.
// 4. `first` — the first realiser in the graph's authored order. Arbitrary,
//    and the page says so rather than dressing it as a recommendation.
//
// ## What this module will not do
//
// - **Walk forever.** The containment has loops (`fixed-point-amplification`
//   needs `state-preparation`, and a state-preparation method can need
//   amplification), so a capability already on the current path is not
//   expanded again, and depth is capped. Both stops are reported on the stage,
//   never silent — a quietly shortened tree reads like a simpler workflow.
// - **Claim two blocks compose.** A path the graph admits is a fact about the
//   contracts; whether the physics lines up is a different claim (roadmap §6).
//   The page shows the walk; it never prints a green "compatible".
import { indexPlannerGraph, type IndexedGraph, type PlannerGraph, type PlannerNode, type PlannerRepeat } from "./graph.ts";
import type { ProblemClass } from "./problems.ts";
import type { Bilingual } from "./types.ts";

export const MAX_STAGE_DEPTH = 3;

export type StageChoice = "reader" | "published" | "preferred" | "first" | "none";

export interface Stage {
  /** Stable address of this position in the tree, e.g. `quantum-linear-solve/state-preparation`. */
  path: string;
  depth: number;
  capability: PlannerNode;
  method: PlannerNode | null;
  choice: StageChoice;
  /** Why the planner started here, when the choice is "preferred" below the root. */
  reason: Bilingual | null;
  /** Every method realising this capability, the chosen one included. */
  alternatives: PlannerNode[];
  /** How often the parent's route runs this step, where the source says. */
  repeat: PlannerRepeat | null;
  children: Stage[];
  stop: "cycle" | "depth" | null;
}

/** Path → method id, for the positions the reader has changed. */
export type MethodChoices = Readonly<Record<string, string>>;

function chooseMethod(
  index: IndexedGraph,
  capabilityId: string,
  path: string,
  choices: MethodChoices,
  parent: PlannerNode | null,
  preferred: readonly string[],
  stepDefault: { method: string; reason: Bilingual } | undefined,
): { method: PlannerNode | null; choice: StageChoice; reason: Bilingual | null } {
  const realisers = index.realisers.get(capabilityId) ?? [];
  const byId = (id: string | undefined) => realisers.find((node) => node.id === id) ?? null;
  const reader = byId(choices[path]);
  if (reader) return { method: reader, choice: "reader", reason: null };
  const published = byId(parent?.via[capabilityId]);
  if (published) return { method: published, choice: "published", reason: null };
  const defaulted = byId(stepDefault?.method);
  if (defaulted && stepDefault) return { method: defaulted, choice: "preferred", reason: stepDefault.reason };
  for (const id of preferred) {
    const node = byId(id);
    if (node) return { method: node, choice: "preferred", reason: null };
  }
  const first = realisers[0] ?? null;
  return { method: first, choice: first ? "first" : "none", reason: null };
}

function buildStage(
  index: IndexedGraph,
  capabilityId: string,
  path: string,
  depth: number,
  onPath: ReadonlySet<string>,
  choices: MethodChoices,
  parent: PlannerNode | null,
  preferred: readonly string[],
  stepDefaults: ProblemClass["stepDefaults"],
): Stage | null {
  const capability = index.byId.get(capabilityId);
  if (!capability || capability.kind !== "capability") return null;
  const stepDefault = depth > 0 ? stepDefaults?.[capabilityId] : undefined;
  const { method, choice, reason } = chooseMethod(index, capabilityId, path, choices, parent, preferred, stepDefault);
  const stage: Stage = {
    path,
    depth,
    capability,
    method,
    choice,
    reason,
    alternatives: index.realisers.get(capabilityId) ?? [],
    repeat: parent?.repeats[capabilityId] ?? null,
    children: [],
    stop: null,
  };
  if (!method || method.steps.length === 0) return stage;
  if (depth >= MAX_STAGE_DEPTH) {
    stage.stop = "depth";
    return stage;
  }
  const nextPath = new Set(onPath).add(capabilityId);
  for (const step of method.steps) {
    if (nextPath.has(step)) {
      // The step is a capability already being expanded above this one. Shown
      // as a stage that stops, so the loop is visible rather than elided.
      const looped = index.byId.get(step);
      if (looped && looped.kind === "capability") {
        stage.children.push({
          path: `${path}/${step}`,
          depth: depth + 1,
          capability: looped,
          method: null,
          choice: "none",
          reason: null,
          alternatives: index.realisers.get(step) ?? [],
          repeat: method.repeats[step] ?? null,
          children: [],
          stop: "cycle",
        });
      }
      continue;
    }
    const child = buildStage(index, step, `${path}/${step}`, depth + 1, nextPath, choices, method, [], stepDefaults);
    if (child) stage.children.push(child);
  }
  return stage;
}

export function assembleWorkflow(graph: PlannerGraph | IndexedGraph, problem: ProblemClass, choices: MethodChoices = {}): Stage | null {
  const index = "byId" in graph ? graph : indexPlannerGraph(graph);
  return buildStage(index, problem.capability, problem.capability, 0, new Set(), choices, null, problem.preferredMethods, problem.stepDefaults);
}

/** Every stage in the tree, parents before children — for lookups and counts. */
export function flattenStages(root: Stage | null): Stage[] {
  if (!root) return [];
  const out: Stage[] = [];
  const walk = (stage: Stage) => {
    out.push(stage);
    for (const child of stage.children) walk(child);
  };
  walk(root);
  return out;
}

/** The method chosen at the first stage realising this capability, if any. */
export function chosenMethodFor(root: Stage | null, capabilityId: string): string | null {
  return flattenStages(root).find((stage) => stage.capability.id === capabilityId)?.method?.id ?? null;
}

// ---------------------------------------------------------------------------
// Compilation: the last block of every workflow

/**
 * Methods built to run without error correction. For these the planner starts
 * the compile stage on NISQ transpilation; for everything else, on
 * fault-tolerant compilation, because their cost lines are counted in Toffoli
 * and T gates, which already assumes an error-corrected machine. The reader can
 * switch either way — this only picks the first choice, and the page says why.
 */
const RUNS_WITHOUT_ERROR_CORRECTION = new Set([
  "variational-ground-state",
  "variational-imaginary-time",
  "qaoa-cost-mixer-alternation",
]);

export const COMPILE_CAPABILITY = "compile-to-device";

export function compileStage(graph: PlannerGraph | IndexedGraph, root: Stage | null, choices: MethodChoices = {}): Stage | null {
  const index = "byId" in graph ? graph : indexPlannerGraph(graph);
  const nisqFirst = root?.method ? RUNS_WITHOUT_ERROR_CORRECTION.has(root.method.id) : false;
  const preferred = nisqFirst ? ["nisq-transpilation"] : ["fault-tolerant-compilation"];
  return buildStage(index, COMPILE_CAPABILITY, COMPILE_CAPABILITY, 0, new Set(), choices, null, preferred, undefined);
}

export function compilesWithoutErrorCorrection(root: Stage | null): boolean {
  return root?.method ? RUNS_WITHOUT_ERROR_CORRECTION.has(root.method.id) : false;
}

File: packages/py/planner/src/leona_planner/assemble.py
"""From a problem to a workflow: the walk down the layer graph, ported from `assemble.ts`.

`apps/web/lib/workflow-planner/assemble.ts` is the source of truth, and its module
comment is the rule this file follows: a workflow is one method chosen at every
capability on a walk down the graph's containment, the default at each position is
the reader's pick, else the parent's own `via`, else a planner preference, else the
first realiser, and the walk stops (visibly) at a capability already on the path and
below `max_stage_depth`. The parity grid assembles every choice the default pipeline
offers in both implementations and fails on any difference.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ._data import Node, PlannerData, Problem

#: `StageChoice` in TS.
READER, PUBLISHED, PREFERRED, FIRST, NONE = "reader", "published", "preferred", "first", "none"

#: The only choice at the root of the compile step the planner makes for a method
#: built to run without error correction, and for everything else (`compileStage`).
NISQ_COMPILE = "nisq-transpilation"
FAULT_TOLERANT_COMPILE = "fault-tolerant-compilation"


@dataclass(eq=False)
class Stage:
    path: str
    depth: int
    capability: Node
    method: Node | None
    choice: str
    reason: dict[str, str] | None
    alternatives: tuple[Node, ...]
    repeat: dict[str, Any] | None
    children: list[Stage] = field(default_factory=list)
    stop: str | None = None


def _choose(
    data: PlannerData,
    capability: str,
    path: str,
    choices: dict[str, str],
    parent: Node | None,
    preferred: tuple[str, ...],
    step_default: dict[str, Any] | None,
) -> tuple[Node | None, str, dict[str, str] | None]:
    realisers = data.realisers.get(capability, ())

    def by_id(wanted: str | None) -> Node | None:
        return next((node for node in realisers if node.id == wanted), None)

    reader = by_id(choices.get(path))
    if reader:
        return reader, READER, None
    published = by_id(parent.via.get(capability) if parent else None)
    if published:
        return published, PUBLISHED, None
    defaulted = by_id(step_default["method"] if step_default else None)
    if defaulted and step_default:
        return defaulted, PREFERRED, step_default["reason"]
    for wanted in preferred:
        node = by_id(wanted)
        if node:
            return node, PREFERRED, None
    first = realisers[0] if realisers else None
    return first, FIRST if first else NONE, None


def _build(
    data: PlannerData,
    capability_id: str,
    path: str,
    depth: int,
    on_path: frozenset[str],
    choices: dict[str, str],
    parent: Node | None,
    preferred: tuple[str, ...],
    step_defaults: dict[str, dict[str, Any]] | None,
) -> Stage | None:
    capability = data.nodes.get(capability_id)
    if capability is None or capability.kind != "capability":
        return None
    step_default = (step_defaults or {}).get(capability_id) if depth > 0 else None
    method, choice, reason = _choose(
        data, capability_id, path, choices, parent, preferred, step_default
    )
    stage = Stage(
        path=path,
        depth=depth,
        capability=capability,
        method=method,
        choice=choice,
        reason=reason,
        alternatives=data.realisers.get(capability_id, ()),
        repeat=parent.repeats.get(capability_id) if parent else None,
    )
    if method is None or not method.steps:
        return stage
    if depth >= data.max_stage_depth:
        stage.stop = "depth"
        return stage
    next_path = on_path | {capability_id}
    for step in method.steps:
        if step in next_path:
            # A capability already being expanded above this one: shown as a stage
            # that stops, so the loop is visible rather than elided.
            looped = data.nodes.get(step)
            if looped is not None and looped.kind == "capability":
                stage.children.append(
                    Stage(
                        path=f"{path}/{step}",
                        depth=depth + 1,
                        capability=looped,
                        method=None,
                        choice=NONE,
                        reason=None,
                        alternatives=data.realisers.get(step, ()),
                        repeat=method.repeats.get(step),
                        stop="cycle",
                    )
                )
            continue
        child = _build(
            data, step, f"{path}/{step}", depth + 1, next_path, choices, method, (), step_defaults
        )
        if child is not None:
            stage.children.append(child)
    return stage


def assemble_workflow(
    data: PlannerData, problem: Problem, choices: dict[str, str] | None = None
) -> Stage | None:
    return _build(
        data,
        problem.capability,
        problem.capability,
        0,
        frozenset(),
        dict(choices or {}),
        None,
        problem.preferred_methods,
        problem.step_defaults,
    )


def compile_stage(
    data: PlannerData, root: Stage | None, choices: dict[str, str] | None = None
) -> Stage | None:
    nisq_first = bool(root and root.method and root.method.id in data.runs_without_error_correction)
    preferred = (NISQ_COMPILE,) if nisq_first else (FAULT_TOLERANT_COMPILE,)
    return _build(
        data,
        data.compile_capability,
        data.compile_capability,
        0,
        frozenset(),
        dict(choices or {}),
        None,
        preferred,
        None,
    )


def flatten(root: Stage | None) -> list[Stage]:
    """Every stage, parents before children."""
    out: list[Stage] = []

    def walk(stage: Stage) -> None:
        out.append(stage)
        for child in stage.children:
            walk(child)

    if root is not None:
        walk(root)
    return out


def chosen_method_for(root: Stage | None, capability_id: str) -> str | None:
    """The method chosen at the first stage realising this capability, if any."""
    for stage in flatten(root):
        if stage.capability.id == capability_id:
            return stage.method.id if stage.method else None
    return None
